tests/test_cases/test_cocotb/test_timing_triggers.py

```diff
@@ -113,7 +116,10 @@ def do_test_afterdelay_in_readonly(dut, delay):
     global exited
     yield RisingEdge(dut.clk)
     yield ReadOnly()
-    yield Timer(delay)
+    with warnings.catch_warnings(record=True) as w:
+        yield Timer(delay)
+        if delay == 0:
+            assert "Timer setup with value 0, which might exhibit undefined behavior in some simulators" in str(w[-1].message)
```

Comment:
Can you move this code to `test_neg_timer` instead? The test that calls this coroutine is marked `expect_fail=...` below, so this line probably never gets executed.

Reply:
It gets executed. Question is: is that a good idea?
The test it now resides in makes the special handling for the simualtor already, the test_beg_timer doesn't, as we just discussed.

Can this cause an unintentional crash?

Reply:
I did the change...let's see how it goes.

Reply:
Ok that doesn't look good:
![grafik](https://user-images.githubusercontent.com/25175069/95488702-2545ae00-0996-11eb-8302-35227195fac4.png)

Reply:
Mhh. Ok I admit this is a bit beyond my knowledge now, why does the CI fail on newer versions of Python?
It seems an Exception is generated which wasn't there before.

I'm apparently not aware of a particular change between versions here. Any hints?

Reply:
It's a CI thing. Any uncaught warnings are turned into exceptions. You need to fix all invocations of `Timer(0)`.

Reply:
Or wrap them in a warnings context manager to avoid the warning

Reply:
But why does this depend on the Python version? That's what I'm not getting.

Reply:
And what other context manager is there for this, aren't we catching the warning already and recording it?
(sorry if I sound stupid, I'm not an expert on the python logging module)

Reply:
It's only run on Python 3.8. I forget exactly why this is and where it is set, @eric-wieser do you remember?. It's bitten me before.

In this case you definitely want to remove all of the `Timer(0)`. They are only used to satisfy the requirement that generators need a `yield` statement in them. The offending tests should have that line removed, and should be converted to `async`/`await` syntax. Sorry, deprecations are never fun :)

Reply:
Ok, but how do we test then? I need to throw that warning at least in the new test for this PR.

And I don't blame any of you, I know how ugly this can get.
At my workplace we also maintain a large Python codebase that always needs to maintain legacy
for at leat 3-4 years, which can be painful^^

Reply:
[This](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONWARNINGS. ) has the answer.

Reply:
Interesting. Didn't know that.

So which approach should be used? Turn it off, demote the warning?

Reply:
By design we run with `exception`, you will need to update the test code so that it isn't tripped. That is the only solution.

Reply:
If that's not clear, the idea is that no other tests should raise the `RuntimeWarning` except this one, which catches it.
So the other tests need to be fixed so that they don't use `Timer(0)`.

Reply:
What @ktbarrett said isn't quite true. There are two possibilities in the tests:

* The test is using `Timer(0)` for no good reason. Change it to `Timer(1)`.
* The test is deliberated testing the undefined behavior of `Timer(0)`. Change it to
  ```python
  with warnings.catch_warnings():
      zero_timer = Timer(0)
  await zero_timer
  ```
  local warning settings override global ones.

Reply:
Sorry, I don't get your point.

How should I test it without firing the warning?
Scrap the test we talked about the last few hours?
You quoted a CLI documentation, which I understand turns wanrings into exceptions, which makes sense for testing in certain scenarios.
But in this case we want the warning, but do not wish to turn of the exception conversion?

Do you wish to make a code suggestion? I'm kind of lost here and did not expect this to turn into such a huge discussion, was kind of just tryin to help out here while waiting for the initial verilator issue to be fixed to continue my project...

Reply:
> How should I test it without firing the warning?

That isn't the test that's failing.

Reply:
> > How should I test it without firing the warning?
> 
> The test you added isn't the test that's failing.

Isn't it? Then I'm not able to read the CI output correctly:
![grafik](https://user-images.githubusercontent.com/25175069/95510877-e8d67a00-09b6-11eb-9939-d1445f0df44a.png)

Reply:
@eric-wieser The only occurrences are case 2. And in those cases they only exist to satisfy the definition of a generator. They are truly pointless statements in the wake of `async`. That's why I recommended removal instead of `Timer(1)`.

Reply:
In my view those 2 tests are identical, except for one doing something illegal, which is the Time(0).
So should we scrap 1 of the tests completely?
![grafik](https://user-images.githubusercontent.com/25175069/95516722-f9d7b900-09bf-11eb-8986-27388b79eeff.png)

Reply:
Ahhhh there is another test. `grep -r` is not a panacea. Personal opinion is the first test should be scrapped. I don't see a reason for us to test a potentially invalid operation in the regression.

Reply:
I would agree, especially now that we emit a warning.

I've said it before: Such stuff is a rabbit-hole of infinite depth, especially with commercial simualtors. They will not look left and right for such stuff in order to maintain competitive performance.

So yes, I would agree to scrap it, but let's wait for some core maintainers.